File: sorting/merge_sort/merge_sort.py
import unittest
from math import floor

def merge_sort(unsorted):
    if len(unsorted) <= 1:
        return unsorted
    left = []
    right = []
    split_index = int(floor(len(unsorted) / 2.0))
    left = unsorted[:split_index]
    # print "left:", left
    right = unsorted[split_index:]
    # print "right:", right
    left = merge_sort(left)
    right = merge_sort(right)
    return merge(left, right)


def merge(left, right):
    result = []
    while len(left) > 0 or len(right) > 0:
        if len(left) > 0 and len(right) > 0:
            if left[0] <= right[0]:
                result.append(left[0])
                left = left[1:]
            else:
                result.append(right[0])
                right = right[1:]
        elif len(left) > 0:
            result.append(left[0])
            left = left[1:]
        elif len(right) > 0:
            result.append(right[0])
            right = right[1:]
    return result

# Testing
class TestMergeSort(unittest.TestCase):
    def test_merge_sort(self):
        self.assertEqual(merge_sort([4,6,2,6,2,1,87,1,7,2,6,2]), [1, 1, 2, 2, 2, 2, 4, 6, 6, 6, 7, 87])
        self.assertEqual(merge_sort([4,3,2,6,8,8,4,3,7,2,6,2]), [2, 2, 2, 3, 3, 4, 4, 6, 6, 7, 8, 8])
        self.assertEqual(merge_sort([3,2,1,4]), [1,2,3,4])
        self.assertEqual(merge_sort([3,2,3,3,1,3,3,4,3,3]), [1,2,3,3,3,3,3,3,3,4])
        self.assertEqual(merge_sort([-1,0,1,2]), [-1,0,1,2])
        self.assertEqual(merge_sort([2,0,1,-1]), [-1,0,1,2])
        self.assertEqual(merge_sort([2,1,0,-1]), [-1,0,1,2])
        self.assertEqual(merge_sort([1,2,1,2,-1,-1]), [-1,-1,1,1,2,2])

    def test_merge(self):
        self.assertEqual(merge([1,4,5], [2,3,6]), [1, 2, 3, 4, 5, 6])

if __name__ == '__main__':
    unittest.main()

File: sorting/merge_sort/merge_sort.rb
require 'minitest/autorun'

# Runtime
# Average: O(n * log(n))
# Best:    O(n * log(n))
# Worst:   O(n * log(n))
#
# Space:   O(n)
def merge_sort(array)
  return array if array.size < 2

  split_index = (array.size / 2.0).floor
  left = array[0...split_index]
  right = array[split_index...array.size]

  merge(merge_sort(left), merge_sort(right))
end

def merge(left, right)
  return right if left.empty?
  return left if right.empty?

  if left[0] <= right[0]
    [ left[0] ] + merge(left[1...left.size], right)
  else
    [ right[0] ] + merge(left, right[1...right.size])
  end
end

class Testing < Minitest::Test
  def test_merge
    assert_equal [1,2,3,4,5,6], merge([1,4,6], [2,3,5])
  end

  def test_merge_sort
    assert_equal [-2,-1,0,1,2], merge_sort([2,1,0,-1,-2])
  end
end
